txnsync/txngroups.go

```diff
@@ -58,3 +227,522 @@ func decodeTransactionGroups(bytes []byte) (txnGroups []transactions.SignedTxGro
+			}
+			groupHash := crypto.HashObj(txGroup)
+			for i := range txns.Transactions {
+				txns.Transactions[i].Txn.Group = groupHash
+			}
+		}
+	}
+}
+
+// deconstructs SignedTxn's into lists of fields and bitmasks
+func deconstructSignedTransactions(stub *txGroupsEncodingStub) {
+	bitmaskLen := (len(stub.SignedTxns)+7)/8 + 1
+	stub.BitmaskAuthAddr = make(bitmask, bitmaskLen)
+	stub.BitmaskLsig = make(bitmask, bitmaskLen)
+	stub.BitmaskMsig = make(bitmask, bitmaskLen)
+	stub.BitmaskSig = make(bitmask, bitmaskLen)
+	for i, txn := range stub.SignedTxns {
+		if !txn.Sig.MsgIsZero() {
+			stub.BitmaskSig.SetBit(i)
+			stub.Sig = append(stub.Sig, txn.Sig)
+			stub.SignedTxns[i].Sig = crypto.Signature{}
+		}
+		if !txn.Msig.MsgIsZero() {
+			stub.BitmaskMsig.SetBit(i)
+			stub.Msig = append(stub.Msig, txn.Msig)
+			stub.SignedTxns[i].Msig = crypto.MultisigSig{}
+		}
+		if !txn.Lsig.MsgIsZero() {
+			stub.BitmaskLsig.SetBit(i)
+			stub.Lsig = append(stub.Lsig, txn.Lsig)
+			stub.SignedTxns[i].Lsig = transactions.LogicSig{}
+		}
+		if !txn.AuthAddr.MsgIsZero() {
+			stub.BitmaskAuthAddr.SetBit(i)
+			stub.AuthAddr = append(stub.AuthAddr, txn.AuthAddr)
+			stub.SignedTxns[i].AuthAddr = basics.Address{}
+		}
+	}
+	stub.BitmaskAuthAddr = trimBitmask(stub.BitmaskAuthAddr)
+	stub.BitmaskLsig = trimBitmask(stub.BitmaskLsig)
+	stub.BitmaskMsig = trimBitmask(stub.BitmaskMsig)
+	stub.BitmaskSig = trimBitmask(stub.BitmaskSig)
+	deconstructTransactions(stub)
+}
+
+func deconstructTransactions(stub *txGroupsEncodingStub) {
+	bitmaskLen := (len(stub.SignedTxns)+7)/8 + 1
+	stub.BitmaskTxType = make(bitmask, bitmaskLen)
+	for i, txn := range stub.SignedTxns {
+		txTypeByte := protocol.TxTypeToByte(txn.Txn.Type)
+		if txTypeByte != 0 {
+			stub.BitmaskTxType.SetBit(i)
+			stub.TxType = append(stub.TxType, txTypeByte)
+		}
+		stub.SignedTxns[i].Txn.Type = ""
+	}
+	stub.BitmaskTxType = trimBitmask(stub.BitmaskTxType)
+	deconstructTxnHeader(stub)
+	deconstructKeyregTxnFields(stub)
+	deconstructPaymentTxnFields(stub)
+	deconstructAssetConfigTxnFields(stub)
+	deconstructAssetTransferTxnFields(stub)
+	deconstructAssetFreezeTxnFields(stub)
+	deconstructApplicationCallTxnFields(stub)
+	deconstructCompactCertTxnFields(stub)
+}
+
+func deconstructTxnHeader(stub *txGroupsEncodingStub) {
+	bitmaskLen := (len(stub.SignedTxns)+7)/8 + 1
+	stub.BitmaskSender = make(bitmask, bitmaskLen)
+	stub.BitmaskFee = make(bitmask, bitmaskLen)
+	stub.BitmaskFirstValid = make(bitmask, bitmaskLen)
+	stub.BitmaskLastValid = make(bitmask, bitmaskLen)
+	stub.BitmaskNote = make(bitmask, bitmaskLen)
+	stub.BitmaskGenesisID = make(bitmask, bitmaskLen)
+	stub.BitmaskGenesisHash = make(bitmask, bitmaskLen)
+	stub.BitmaskLease = make(bitmask, bitmaskLen)
+	stub.BitmaskRekeyTo = make(bitmask, bitmaskLen)
+	for i, txn := range stub.SignedTxns {
+		if !txn.Txn.Sender.MsgIsZero() {
+			stub.BitmaskSender.SetBit(i)
+			stub.Sender = append(stub.Sender, txn.Txn.Sender)
```

Comment:
using a single byte array ([]byte) instead of array of arrays ([][]byte) would result in more efficient encoding. This is relatively easy to do for bytes array of fixed size ( such as `Address` = 32 bytes )